src/secure_binary.rs

```diff
@@ -716,18 +716,36 @@ pub fn show(filename: &str) -> Result<Vec<u8>> {
         let (i, digest_hmac) = take::<_, _, ()>(32u8)(i)?;
         let (i, keyblob) = Keyblob::from_bytes(i)?;
         let (i, certificate_block_header) = FullCertificateBlockHeader::from_bytes(i)?;
-        let (i, certificate_length) = le_u32::<_, ()>(i).unwrap();
-        let (i, certificate_data) = take::<_, _, ()>(certificate_length)(i)?;
-
-        let unpadded_signed_data_length = 16 * (6 + 2 + 5 + 2) + 4 + certificate_length + 128;
+        let (mut i, mut certificates): (_, Vec<X509Certificate<'_>>) = (i, Vec::new());
+        for _ in 0..certificate_block_header.certificate_count {
+            let (itmp, certificate_length) = le_u32::<_, ()>(i).unwrap();
+            let (itmp, certificate_data) = take::<_, _, ()>(certificate_length)(itmp)?;
+            println!("certificate length: {}", certificate_length);
+            println!("end of cert data: {:>16x}", hex_str!(&certificate_data));
+            i = itmp;
+            match X509Certificate::from_der(certificate_data) {
+                Ok((rem, cert)) => {
+                    println!("Remainder: {}", hex_str!(rem));
+                    assert_eq!(
+                        cert.tbs_certificate.version,
+                        x509_parser::x509::X509Version::V3
+                    );
+                    if certificates.len() != 0 {
+                        assert!(cert
+                            .verify_signature(Some(
+                                certificates[certificates.len() - 1].public_key()
+                            ))
+                            .is_ok());
+                    } else {
+                        assert!(cert.verify_signature(None).is_ok());
+                    }
```

Comment:
I think this would be easier to read if you first determine the public key to check and the perform the signature verification only once.  Maybe something like this:

```
    let public_key = certificates.last().map(|c| c.public_key());
    assert!(cert.verify_signature(public_key).is_ok());
```

Reply:
That's a lot nicer indeed. TIL about `Vec::last`